### ISSUE TITLE
Add duckdb plugin.

### ISSUE THREAD

**Issue by lambertsbennett** [nushell/nushell#10920]
### Related problem

I recently came across Nushell and it looks great! I'm new to Rust, but I've been using Duckdb (https://duckdb.org/) a lot. Particularly its useful to query parquet files rapidly. I read the contributing guide and I am not sure whether you can willy nilly add plugins or whether a core contributor should give the go ahead first?

### Describe the solution you'd like

Add a plugin nu-duckdb that allows you to quickly use duckdb functionality in nushell. Although I'm pretty new to Rust I think it should be pretty straightforward to implement.

**Comment by fdncred** [comment]
People should feel free to create their own repo with any plugin they want. I have a bunch in my personal repo. I typically use `nu_plugin_from_parquet` to play with parquet files, although it's not 100% complete.

If you write this plugin, I'll definitely add it to our list in the awesome-nu repo https://github.com/nushell/awesome-nu.

**Comment by lambertsbennett** [comment]
Ok great! I'll work on this then and add a PR to https://github.com/nushell/awesome-nu when I'm done!

**Comment by fdncred** [comment]
Cool. I'm excited to play with it. Ping me when you're ready for a tester.

**Comment by fdncred** [comment]
oh, wait a second. I'm totally ignorant about duckdb. It sounds like it's an in-memory db. If that's correct, and it works the same as sqlite's in-memory db, I'm not sure it'll work in a plugin. The reason for this is that nushell invoked plugins and they run, execute, and exit. So, if an in-memory db is loaded in a plugin, it could initialize and do whateer was passed into it with that first invocation, but when the plugin exits, all memory for the plugin is gone.

one thing I've been experimenting with is an in-memory db for sqlite. you can see how I do it in this PR https://github.com/nushell/nushell/pull/10691

**Comment by lambertsbennett** [comment]
Ah ok! Thanks for the heads up before I went too crazy on this! I'll look at the PR and see what I can do also in the direction of integrating duckdb.

**Comment by fdncred** [comment]
Feel free to take my `$db` PR and make it use duckdb with your selling points. We're getting closer to wanting to have an in-memory db to test, so it would be cool to pick the right one.

**Comment by lambertsbennett** [comment]
OK I looked at the PR and being new to both Rust and nushell this is going to take me a while, but I think in the end DuckDB is a better choice for an in-memory database, so I will work on adapting your work to use DuckDB instead.

**Comment by fdncred** [comment]
@lambertsbennett I'm not sure you can see this. You may need a login to hackmd but I documented/brainstormed a set of commands we may use for an in-memory database. https://hackmd.io/oaC4l4yaRma1MGNv37ydDQ

This would probably remove the `$db` aspect of my PR and move everything to those commands. Just wanted to share because it's an alternate route of doing something like this.

**Comment by lambertsbennett** [comment]
@fdncred I can see the above link thanks! It looks good to me, I was thinking also that having an init option would be good because for duckdb you can choose completely in memory or file backed. If it's alright with you I'll take a crack at implementing some of the features with duckdb. I think a lot of the basic features should be pretty easy. The duckdb-rs crate tries to emulate rusqlite so a lot of stuff would be incredibly similar.

**Comment by fdncred** [comment]
@lambertsbennett yes, please do. sounds like an interesting experiment.

**Comment by lambertsbennett** [comment]
@fdncred So I've been working on this and there are a couple things that I'm struggling with that I'm not sure if you have some ideas? Most things make good sense and I read your comment about not containing the connection within the database struct and thought to myself that it would be easier if we did in fact just store the connection. But I've come up against a wall of the connection not being serializable. This in general is a problem when implementing CustomValue for the Database struct. I figured that we could use something like Mutex to make sure there are no problems with multiple connections to the database, but perhaps I'm completely on the wrong track.
